backend/migrating/v2/management/commands/migrate_to_v2.py

```diff
@@ -0,0 +1,442 @@
+import json
+    import logging
+    import os
+    import time
+    from contextlib import contextmanager
+    from typing import Any, Generator, Optional
+
+    import psycopg2
+    from django.conf import settings
+    from django.core.management.base import BaseCommand
+    from migrating.v2.constants import V2
+    from migrating.v2.unstract_migrations import UnstractMigration
+    from psycopg2.extensions import connection, cursor
+
+    logger = logging.getLogger(__name__)
+
+
+    class DataMigrator:
+        def __init__(self, src_db_config, dest_db_config, v2_schema, batch_size=1000):
+            self.src_db_config = src_db_config
+            self.dest_db_config = dest_db_config
+            self.batch_size = batch_size
+            self.v2_schema = v2_schema
+            self.migration_tracking_table = f"{self.v2_schema}.migration_tracking"
+
+        @contextmanager
+        def _db_connect_and_cursor(self, db_config: dict) -> Generator[tuple[connection, cursor], None, None]:
+            """A context manager to manage database connection and cursor.
+
+            Args:
+                db_config (dict): Database configuration dictionary.
+
+            Yields:
+                tuple[connection, cursor]: A tuple containing the connection and cursor.
+            """
+            conn: connection = psycopg2.connect(**db_config)
+            try:
+                with conn.cursor() as cur:  # Manage cursor within the same context
+                    yield conn, cur
+            except psycopg2.Error as e:
+                conn.rollback()  # Rollback on error
+                logger.error(f"Database operation error: {e}")
+                raise
+            finally:
+                conn.close()
+
+        def _create_tracking_table_if_not_exists(self):
+            """Ensures that the migration tracking table exists in the
+            destination database.
+
+            Creates the table if it does not exist.
+            """
+            with self._db_connect_and_cursor(self.dest_db_config) as (conn, cur):
+                try:
+                    cur.execute(
+                        f"""
+                        CREATE TABLE IF NOT EXISTS {self.migration_tracking_table} (
+                            id SERIAL PRIMARY KEY,
+                            migration_name VARCHAR(255) UNIQUE NOT NULL,
+                            applied_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
+                        );
+                    """
+                    )
+                    conn.commit()
+                    logger.info("Migration tracking table ensured.")
+                except psycopg2.Error as e:
+                    logger.error(f"Error creating migration tracking table: {e}")
```

Comment:
@muhammad-ali-e since you have caught the error here and not re-raised it - it will not be caught and rolled back in the context manager. Is this intentional?

Reply:
Good catch

Reply:
updated